src/msw/joystick.cpp

```diff
@@ -37,8 +37,127 @@
+   if ((x & 0x0000000F) == 0) {n = n + 4; x = x >> 4;}
+   if ((x & 0x00000003) == 0) {n = n + 2; x = x >> 2;}
+   return n - (x & 1);
+}
+#endif
+
+enum {
+    wxJS_AXIS_X = 0,
+    wxJS_AXIS_Y,
+    wxJS_AXIS_Z,
+    wxJS_AXIS_RUDDER,
+    wxJS_AXIS_U,
+    wxJS_AXIS_V,
+
+    wxJS_AXIS_MAX = 32767,
+    wxJS_AXIS_MIN = -32767,
+    wxJS_MAX_AXES = 6,      // WinMM supports up to 6 axes.
+    wxJS_MAX_BUTTONS = 32,  // WinMM supports up to 32 buttons.
+};
+
 wxIMPLEMENT_DYNAMIC_CLASS(wxJoystick, wxObject);
 
+////////////////////////////////////////////////////////////////////////////
+// Background thread for reading the joystick device
+////////////////////////////////////////////////////////////////////////////
+
+class wxJoystickThread : public wxThread
+{
+public:
+    wxJoystickThread(int joystick);
+    void* Entry() wxOVERRIDE;
+
+private:
+    void      SendEvent(wxEventType type, long ts, int change = 0);
+    int       m_joystick;
+    int       m_buttons;
+    wxWindow* m_catchwin;
+    int       m_polling;
+    JOYINFO   m_joyInfo;
+    JOYINFO   m_lastJoyInfo;
+
+    friend class wxJoystick;
+};
+
+
+wxJoystickThread::wxJoystickThread(int joystick)
+    : m_joystick(joystick),
+      m_buttons(0),
+      m_catchwin(NULL),
+      m_polling(0),
+      m_joyInfo(),
+      m_lastJoyInfo()
+{
+}
+
+void wxJoystickThread::SendEvent(wxEventType type, long ts, int change)
+{
+    wxJoystickEvent jwx_event(type, m_buttons, m_joystick, change);
+
+    jwx_event.SetTimestamp(ts);
+    jwx_event.SetPosition(wxPoint(m_joyInfo.wXpos, m_joyInfo.wYpos));
+    jwx_event.SetZPosition(m_joyInfo.wZpos);
+    jwx_event.SetEventObject(m_catchwin);
+
+    if (m_catchwin)
+        m_catchwin->GetEventHandler()->AddPendingEvent(jwx_event);
+}
+
+void* wxJoystickThread::Entry()
+{
+    joyGetPos(m_joystick, &m_lastJoyInfo);
+
+    while (true)
+    {
+        if (TestDestroy())
+            break;
+
+        this->Sleep(m_polling);
+        DWORD ts = GetTickCount();
+
+        joyGetPos(m_joystick, &m_joyInfo);
+        m_buttons = m_joyInfo.wButtons;
+        DWORD delta = m_buttons ^ m_lastJoyInfo.wButtons;
+        DWORD deltaUp = delta & !m_buttons;
+        DWORD deltaDown = delta & m_buttons;
+
+        // Use count trailing zeros to determine which button changed.
+        // Was using JOYINFOEX.dwButtons, because the docs state this is
+        // "Current button number that is pressed.", but it turns out
+        // it is the *total* number of buttons pressed.
+        if (deltaUp)
+            SendEvent(wxEVT_JOY_BUTTON_UP, ts, ctz(deltaUp)+1);
+        if(deltaDown)
+            SendEvent(wxEVT_JOY_BUTTON_DOWN, ts, ctz(deltaDown)+1);
+
+        if ((m_joyInfo.wXpos != m_lastJoyInfo.wXpos) ||
+            (m_joyInfo.wYpos != m_lastJoyInfo.wYpos) ||
+            (m_joyInfo.wZpos != m_lastJoyInfo.wZpos)   )
+        {
+            SendEvent(wxEVT_JOY_MOVE, ts);
+        }
+
+        memcpy (&m_lastJoyInfo, &m_joyInfo, sizeof(JOYINFO));
```

Comment:
This is unnecessary, `m_lastJoyInfo = m_joyInfo` is more readable and works just as well in C++.

Reply:
Oops. Habit from usually working C & python, I guess. Fixed.